internal/logic/logic_test.go

```diff
@@ -51,73 +58,217 @@ func Test_transactionManagementServiceLogic_HealthCheck(t *testing.T) {
+						CurrentPage: 1,
+						NextPage:    -1,
+						TotalPage:   1,
+					}, res.Pagination)
+				}
 			},
-			give: &model.PingRequest{
-				Data: "ping",
+		},
+		{
+			name:        "Success :: AccDetails:: count_offset>limit",
+			credentials: "123",
+			setup: func() (datasource.DataSourceI, string) {
+				mockDs := mock.NewMockDataSourceI(mockCtrl)
+				var trans []model.Transaction
+				trans = append(trans, model.Transaction{UserId: "123", AccountNumber: 1})
+				mockDs.EXPECT().Get(map[string]interface{}{"user_id": "123"}, 5, 0).Times(1).Return(trans, 100, nil)
+				return mockDs, ""
 			},
-			want: &respModel.Response{
-				Status:  http.StatusOK,
-				Message: "Pong",
-				Data: &model.DsResponse{
-					Data: "pong",
-				},
+			want: func(resp *respModel.Response) {
+				var users = model.GetTransaction{AccountNumber: 1}
+				temp := respModel.Response{
+					Status:  http.StatusOK,
+					Message: "SUCCESS",
+					Data:    users,
+				}
+				res, ok := resp.Data.(model.PaginatedResponse)
+				if !ok {
+					t.Log("fail")
+					t.Fail()
+				}
+				if !reflect.DeepEqual(&resp.Status, &temp.Status) {
+					t.Errorf("Want: %v, Got: %v", &temp.Status, &resp.Status)
+				}
+				if !reflect.DeepEqual(&resp.Message, &temp.Message) {
+					t.Errorf("Want: %v, Got: %v", &temp.Message, &resp.Message)
+				}
+				if !reflect.DeepEqual(res.Response[0].AccountNumber, 1) {
+					t.Errorf("Want: %v, Got: %v", 1, res.Response[0].AccountNumber)
+				}
+				if !reflect.DeepEqual(res.Pagination, model.Paginate{
+					CurrentPage: 1,
+					NextPage:    2,
+					TotalPage:   20,
+				}) {
+					t.Errorf("Want: %v, Got: %v", model.Paginate{
+						CurrentPage: 1,
+						NextPage:    -1,
+						TotalPage:   1,
+					}, res.Pagination)
+				}
 			},
 		},
 		{
-			name: "Failure::datasource error",
-			setup: func() datasource.DataSource {
-				mockDs := mock.NewMockDataSource(mockCtrl)
+			name:        "Failure :: AccDetails :: db err",
+			credentials: "123",
+			setup: func() (datasource.DataSourceI, string) {
+				mockDs := mock.NewMockDataSourceI(mockCtrl)
+				mockDs.EXPECT().Get(map[string]interface{}{"user_id": "123"}, 5, 0).Times(1).Return(nil, 0, errors.New("error"))
+				return mockDs, ""
+			},
+			want: func(resp *respModel.Response) {
+				temp := respModel.Response{
+					Status:  http.StatusInternalServerError,
+					Message: codes.GetErr(codes.ErrGetTransaction),
+					Data:    nil,
+				}
+				if !reflect.DeepEqual(resp, &temp) {
+					t.Errorf("Want: %v, Got: %v", &temp, resp)
+				}
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rec := NewTransactionManagementServiceLogic(tt.setup())
 
-				mockDs.EXPECT().Ping(&model.PingDs{
-					Data: "ping",
-				}).Times(1).
-					Return(nil, errors.New("ds down"))
+			got := rec.GetTransactions(tt.credentials, 5, 1)
 
-				return mockDs
+			tt.want(got)
+		})
+	}
+}
+
+func TestTransactionManagementServiceLogic_NewTransaction(t *testing.T) {
```

Comment:
I have added a test server to mock but there is a race condition in the code due to which I am not able to  record or dump any data from test server as the test completes becouse request is made inside go routine
I was able to add test case to validate all the fileds of data inside test server but it only works only if I add time.sleep in logic just after go routine so how to test in case of these race condition?

Reply:
Are these implemented?

Reply:
Test case can be added for this flow: https://app.codecov.io/gh/vatsal278/TransactionManagementService/pull/4#ebea564d94e0d939ccfa29b6ab98b11a-R113